rclrs/src/node/client.rs

```diff
@@ -0,0 +1,253 @@
+            rcl_client_mtx: Mutex::new(rcl_client),
+            rcl_node_mtx: node.rcl_node_mtx.clone(),
+        });
+
+        Ok(Self {
+            handle,
+            requests: Mutex::new(HashMap::new()),
+            futures: Arc::new(Mutex::new(
+                HashMap::<i64, oneshot::Sender<T::Response>>::new(),
+            )),
+        })
+    }
+
+    /// Send a requests with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub fn async_send_request_with_callback<'a, M: MessageCow<'a, T::Request>, F>(
+        &self,
+        message: M,
+        callback: F,
+    ) -> Result<(), RclrsError>
+    where
+        F: FnOnce(&T::Response) + 'static + Send,
+    {
+        let rmw_message = T::Request::into_rmw_message(message.into_cow());
+        let mut sequence_number = -1;
+        let ret = unsafe {
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        };
+        let requests = &mut *self.requests.lock();
+        requests.insert(sequence_number, Box::new(callback));
+        ret.ok()
+    }
+
+    /// Send a request with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub async fn call_async<'a, R: MessageCow<'a, T::Request>>(
+        &self,
+        request: R,
+    ) -> Result<T::Response, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service + 'static,
+    {
+        let rmw_message = T::Request::into_rmw_message(request.into_cow());
+        let mut sequence_number = -1;
+        let ret = unsafe {
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
```

Comment:
I think this const-to-mut cast can be avoided by making the `rmw_message` variable `mut` and using `as_mut()` instead of `as_ref()`. Same for the service.

Reply:
Yes, definitely.

Reply:
Yes, `Cow` doesn't have an `AsMut` impl. I think it's fine to leave as it is, since the message should not be modified by `rcl_send_request` anyway.